sam/assembly_test.go

```diff
@@ -0,0 +1,163 @@
+package sam
+
+import (
+	"fmt"
+	"github.com/vertgenlab/gonomics/dna"
+	"testing"
+)
+
+var DiploidBaseCallFromPileTests = []struct {
+	P        Pile
+	RefBase  dna.Base
+	Delta    float64
+	Gamma    float64
+	Epsilon  float64
+	Expected DiploidBase
+}{
+	{P: Pile{CountF: [13]int{16, 14, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0}}, //good coverage heterozygous double mutation
+		RefBase:  dna.G,
+		Delta:    0.01,
+		Gamma:    3,
+		Epsilon:  0.01,
+		Expected: AC},
+	{P: Pile{CountF: [13]int{4, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0}}, //little data, only confident enough to call one mutation
+		RefBase:  dna.G,
+		Delta:    0.01,
+		Gamma:    3,
+		Epsilon:  0.01,
+		Expected: AG},
+	{P: Pile{CountF: [13]int{4, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0}}, //changing expected divergence rate changes pile outcome
+		RefBase:  dna.G,
+		Delta:    0.5,
+		Gamma:    3,
+		Epsilon:  0.01,
+		Expected: AT},
+	{P: Pile{CountF: [13]int{4, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0}}, //changing expected error rate changes pile outcome.
+		RefBase:  dna.G,
+		Delta:    0.01,
+		Gamma:    3,
+		Epsilon:  0.0001,
+		Expected: AT},
+	{P: Pile{CountF: [13]int{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}}, //test empty pile, returns homozygous reference
+		RefBase:  dna.G,
+		Delta:    0.01,
+		Gamma:    3,
+		Epsilon:  0.01,
+		Expected: GG},
+	{P: Pile{CountF: [13]int{16, 450, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0}}, //stress test, huge pile
+		RefBase:  dna.G,
+		Delta:    0.01,
+		Gamma:    3,
+		Epsilon:  0.01,
+		Expected: CC},
+	{P: Pile{CountF: [13]int{16, 14, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0}}, //reference N returns NN.
+		RefBase:  dna.N,
+		Delta:    0.01,
+		Gamma:    3,
+		Epsilon:  0.01,
+		Expected: NN},
+	{P: Pile{CountF: [13]int{16, 1, 0, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0}}, //messy pile, high error rate
+		RefBase:  dna.C,
+		Delta:    0.1,
+		Gamma:    3,
+		Epsilon:  0.01,
+		Expected: AT},
+}
+
+func TestDiploidBaseCallFromPile(t *testing.T) {
+	var actual DiploidBase
+	var priorCache [][]float64
+	var heterozycousCache = make([][]float64, 0)
```

Comment:
Where are the caches actually being defined? I am looking through the functions and I can't find where they are being built???

Reply:
Right, so for the tests, we aren't using caches and calculating everything by hand. The plan is that the cmd (to be called samAssembler) will make caches and pass them to the package functions. You can see the functions to initialize the caches at the end of the file

Reply:
I see now, so you check to see if the cache is available else calculate by hand. You should consider making the cache building functions public in assembly.go assuming the caches will be broadly applicable. Having DiploidBaseCallFromPile be a public library function, but making the functions to build the caches for it private makes it pretty hard for other people to use DiploidBaseCallFromPile.

Reply:
Ordinarily I would, but for the likelihood, I think cache building will be trivial: we'll make a 100 by 100 [][]float64, all initialized to zero. We expect sparse usage of the whole cache space (most piles should have the same approximate coverage, and generally be pretty neat either het or homozygous), so we won't precalculate everything, and fill the cache as we progress through the program